File: dts/lib/models/agreement_model.dart
import 'dart:convert';

class DescriptionItem {
  final String description;
  final double quantity;
  final double rate;
  final double subTotal;

  DescriptionItem({
    required this.description,
    required this.quantity,
    required this.rate,
    required this.subTotal,
  });

  factory DescriptionItem.fromJson(Map<String, dynamic> json) {
    final qty = (json['quantity'] as num?)?.toDouble() ?? 0.0;
    final r = (json['rate'] as num?)?.toDouble() ?? 0.0;
    return DescriptionItem(
      description: json['description'] ?? '',
      quantity: qty,
      rate: r,
      subTotal: (json['subTotal'] as num?)?.toDouble() ?? (qty * r),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'description': description,
      'quantity': quantity,
      'rate': rate,
      'subTotal': subTotal,
    };
  }
}

class AgreementModel {
  final String? id;
  final String documentType; // 'Agreement' or 'Quotation'
  final String? offerNumber;
  final DateTime date;
  final String customerName;
  final String completeAddress;
  final String contactPerson;
  final String mobileNumber;
  final List<DescriptionItem> descriptionItems;
  final bool gstRequired;
  final double gstPercentage;
  final double totalBeforeGST;
  final double gstAmount;
  final double grandTotal;
  final String? amountInWords;
  final String technicianSignatureUrl;
  final String? customerSignatureUrl;
  final String? termsAndConditions;
  final String? paymentTerms;
  final String? offerValidity;
  final String? notes;
  final String? footerText;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  AgreementModel({
    this.id,
    required this.documentType,
    this.offerNumber,
    required this.date,
    required this.customerName,
    required this.completeAddress,
    required this.contactPerson,
    required this.mobileNumber,
    required this.descriptionItems,
    required this.gstRequired,
    required this.gstPercentage,
    required this.totalBeforeGST,
    required this.gstAmount,
    required this.grandTotal,
    this.amountInWords,
    required this.technicianSignatureUrl,
    this.customerSignatureUrl,
    this.termsAndConditions,
    this.paymentTerms,
    this.offerValidity,
    this.notes,
    this.footerText,
    this.createdAt,
    this.updatedAt,
  });

  factory AgreementModel.fromJson(Map<String, dynamic> json) {
    final itemsRaw = json['descriptionItems'] as List?;
    final items = itemsRaw != null
        ? itemsRaw.map((e) => DescriptionItem.fromJson(e as Map<String, dynamic>)).toList()
        : <DescriptionItem>[];

    return AgreementModel(
      id: json['_id'] ?? json['id'],
      documentType: json['documentType'] ?? 'Agreement',
      offerNumber: json['offerNumber'],
      date: json['date'] != null 
          ? DateTime.tryParse(json['date'].toString()) ?? DateTime.now() 
          : DateTime.now(),
      customerName: json['customerName'] ?? '',
      completeAddress: json['completeAddress'] ?? '',
      contactPerson: json['contactPerson'] ?? '',
      mobileNumber: json['mobileNumber'] ?? '',
      descriptionItems: items,
      gstRequired: json['gstRequired'] == true || json['gstRequired'] == 'true',
      gstPercentage: (json['gstPercentage'] as num?)?.toDouble() ?? 0.0,
      totalBeforeGST: (json['totalBeforeGST'] as num?)?.toDouble() ?? 0.0,
      gstAmount: (json['gstAmount'] as num?)?.toDouble() ?? 0.0,
      grandTotal: (json['grandTotal'] as num?)?.toDouble() ?? 0.0,
      amountInWords: json['amountInWords'],
      technicianSignatureUrl: json['technicianSignatureUrl'] ?? '',
      customerSignatureUrl: json['customerSignatureUrl'],
      termsAndConditions: json['termsAndConditions'],
      paymentTerms: json['paymentTerms'],
      offerValidity: json['offerValidity'],
      notes: json['notes'],
      footerText: json['footerText'],
      createdAt: json['createdAt'] != null ? DateTime.tryParse(json['createdAt'].toString()) : null,
      updatedAt: json['updatedAt'] != null ? DateTime.tryParse(json['updatedAt'].toString()) : null,
    );
  }

  Map<String, dynamic> toJson({bool flat = false}) {
    final map = <String, dynamic>{
      'documentType': documentType,
      'date': date.toIso8601String(),
      'customerName': customerName,
      'completeAddress': completeAddress,
      'contactPerson': contactPerson,
      'mobileNumber': mobileNumber,
      'descriptionItems': descriptionItems.map((e) => e.toJson()).toList(),
      'gstRequired': gstRequired,
      'gstPercentage': gstPercentage,
      'totalBeforeGST': totalBeforeGST,
      'gstAmount': gstAmount,
      'grandTotal': grandTotal,
      'technicianSignatureUrl': technicianSignatureUrl,
    };

    if (customerSignatureUrl != null) {
      map['customerSignatureUrl'] = customerSignatureUrl;
    }
    if (termsAndConditions != null) {
      map['termsAndConditions'] = termsAndConditions;
    }
    if (paymentTerms != null) {
      map['paymentTerms'] = paymentTerms;
    }
    if (offerValidity != null) {
      map['offerValidity'] = offerValidity;
    }
    if (notes != null) {
      map['notes'] = notes;
    }
    if (footerText != null) {
      map['footerText'] = footerText;
    }
    if (id != null) {
      map['id'] = id;
    }

    return map;
  }

  String toJsonString() {
    return jsonEncode(toJson());
  }
}

File: backend/src/models/agreement.model.js
import mongoose from 'mongoose';

const descriptionItemSchema = new mongoose.Schema(
  {
    description: {
      type: String,
      trim: true,
      required: true,
    },
    quantity: {
      type: Number,
      required: true,
      min: 0.0001,
    },
    rate: {
      type: Number,
      required: true,
      min: 0.0001,
    },
    subTotal: {
      type: Number,
      default: 0,
    },
  },
  { _id: false }
);

const agreementSchema = new mongoose.Schema(
  {
    documentType: {
      type: String,
      enum: ['Agreement', 'Quotation'],
      required: true,
      index: true,
    },
    offerNumber: {
      type: String,
      unique: true,
      trim: true,
      index: true,
    },
    date: {
      type: Date,
      required: true,
      default: Date.now,
      index: true,
    },
    customerName: {
      type: String,
      trim: true,
      index: true,
    },
    completeAddress: {
      type: String,
      trim: true,
    },
    contactPerson: {
      type: String,
      trim: true,
    },
    mobileNumber: {
      type: String,
      trim: true,
      index: true,
    },
    descriptionItems: {
      type: [descriptionItemSchema],
      default: [],
    },
    gstRequired: {
      type: Boolean,
      default: false,
    },
    gstPercentage: {
      type: Number,
      min: 0,
      max: 100,
      default: 0,
    },
    totalBeforeGST: {
      type: Number,
      default: 0,
    },
    gstAmount: {
      type: Number,
      default: 0,
    },
    grandTotal: {
      type: Number,
      default: 0,
    },
    amountInWords: {
      type: String,
      trim: true,
    },
    technicianSignatureUrl: {
      type: String,
      trim: true,
      required: true,
      default: 'https://res.cloudinary.com/dy5gs2egc/image/upload/v1782710059/efsr/signatures/i1ijhzyhgkmeig7v7cad.png',
    },
    customerSignatureUrl: {
      type: String,
      trim: true,
    },
    termsAndConditions: {
      type: String,
      trim: true,
      default: 'All services are subject to the agreed AMC scope, scheduled maintenance, and site accessibility. Any additional work outside the agreed scope will be billed separately.',
    },
    paymentTerms: {
      type: String,
      trim: true,
      default: 'Payment is due within 15 days from the date of invoice.',
    },
    offerValidity: {
      type: String,
      trim: true,
      default: 'This offer is valid for 15 days from the date of issue.',
    },
    notes: {
      type: String,
      trim: true,
      default: 'Please confirm the scope and service commencement date at the earliest.',
    },
    footerText: {
      type: String,
      trim: true,
      default: 'Thank you for choosing GPS Technical Services.',
    },
    numberOfFreeVisits: {
      type: Number,
      default: 6,
    },
    status: {
      type: String,
      enum: ['draft', 'submitted'],
      default: 'submitted',
      index: true,
    },
  },
  {
    timestamps: true,
  }
);

agreementSchema.index({ offerNumber: 'text', customerName: 'text', mobileNumber: 'text', documentType: 'text' });

const Agreement = mongoose.model('Agreement', agreementSchema);

export default Agreement;
